File: typescript/src/answers.ts
import type { PendingDecisionView, PreviousTurn } from "./types.js";

/** Business-level answer; identifiers are resolved from the pending candidate. */
export interface AnswerInput {
  readonly userText?: string;
  readonly candidateRank?: number;
  readonly choiceLabels?: readonly string[];
  readonly freeText?: string;
  readonly outcome?: PreviousTurn["outcome"];
  readonly assistantText?: string;
}

export function answerTurn(pending: PendingDecisionView | null, answer: AnswerInput): PreviousTurn {
  if (!pending) throw new Error("No pending question. Call next() first.");
  const matches = pending.candidates.filter((c) => c.rank === (answer.candidateRank ?? 1));
  const candidate = matches[0];
  if (matches.length !== 1 || !candidate) throw new Error("Unknown candidate rank");
  let structured: PreviousTurn["structured_answer"];
  if (answer.choiceLabels !== undefined || answer.freeText !== undefined) {
    if (candidate.type === "open") throw new Error("Open questions accept userText, not choices");
    const labels = answer.choiceLabels ?? [];
    if (new Set(labels).size !== labels.length) throw new Error("Duplicate choice labels");
    const ids = labels.map((label) => {
      const choices = candidate.choices.filter((c) => c.label === label);
      if (choices.length !== 1 || !choices[0]) throw new Error("Unknown or ambiguous choice label");
      return choices[0].choice_id;
    });
    if (
      ids.length > 1 &&
      (candidate.type === "single_choice" ||
        (candidate.type === "semi_open" && candidate.selection_mode !== "multiple"))
    ) {
      throw new Error("This question accepts only one choice");
    }
    if (answer.freeText !== undefined && candidate.type !== "semi_open") {
      throw new Error("Free text is only allowed for semi-open choices");
    }
    if (!ids.length)
      throw new Error("Select at least one choice; for an unlisted answer send userText");
    structured = {
      choice_ids: ids,
      ...(answer.freeText !== undefined ? { free_text: answer.freeText } : {}),
    };
  }
  if (answer.userText === undefined && !structured && answer.outcome === undefined) {
    throw new Error("Provide an answer, choices, or an explicit outcome");
  }
  if (answer.userText !== undefined && !answer.userText.trim()) throw new Error("Empty answer");
  return {
    decision_id: pending.decision_id,
    question_id: candidate.question_id,
    assistant_text: answer.assistantText ?? candidate.text,
    ...(answer.userText !== undefined ? { user_text: answer.userText } : {}),
    ...(answer.outcome !== undefined ? { outcome: answer.outcome } : {}),
    ...(structured ? { structured_answer: structured } : {}),
  };
}

File: python/src/zelinqa/answers.py
"""Resolve a business answer against the pending decision, without an HTTP call."""

from collections.abc import Sequence

from .models import PendingDecisionView, PreviousTurn, QuestionOutcome, StructuredAnswer


def answer_turn(
    pending: PendingDecisionView | None,
    *,
    user_text: str | None = None,
    candidate_rank: int = 1,
    choice_labels: Sequence[str] | None = None,
    free_text: str | None = None,
    outcome: QuestionOutcome | None = None,
    assistant_text: str | None = None,
) -> PreviousTurn:
    if pending is None:
        raise ValueError("No pending question. Call next() first.")
    candidates = [c for c in pending.candidates if c.rank == candidate_rank]
    if len(candidates) != 1:
        raise ValueError("candidate_rank must identify one pending candidate")
    candidate = candidates[0]
    structured = None
    if choice_labels is not None or free_text is not None:
        if candidate.type == "open":
            raise ValueError("Open questions accept user_text, not choices")
        labels = list(choice_labels or [])
        if len(set(labels)) != len(labels):
            raise ValueError("Duplicate choice labels")
        ids = []
        for label in labels:
            matches = [c.choice_id for c in candidate.choices if c.label == label]
            if len(matches) != 1:
                raise ValueError("Unknown or ambiguous choice label; use exact displayed labels")
            ids.append(matches[0])
        if len(ids) > 1 and (
            candidate.type == "single_choice"
            or (candidate.type == "semi_open" and candidate.selection_mode != "multiple")
        ):
            raise ValueError("This question accepts only one choice")
        if free_text is not None and candidate.type != "semi_open":
            raise ValueError("Free text is only allowed for semi-open choices")
        if not ids:
            raise ValueError(
                "Select at least one choice; for an unlisted answer, send user_text instead"
            )
        structured = StructuredAnswer(choice_ids=ids, free_text=free_text)
    if user_text is None and structured is None and outcome is None:
        raise ValueError("Provide an answer, choices, or an explicit outcome")
    return PreviousTurn(
        decision_id=pending.decision_id,
        question_id=candidate.question_id,
        assistant_text=assistant_text or candidate.text,
        user_text=user_text,
        structured_answer=structured,
        outcome=outcome,
    )
